File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    var answer = 0;
    
    let obj1 = {};
    let obj2 = {};
    let cnt1 =0;
    let cnt2 =0;
    
    for(const item of topping) {
        if(obj2[item]) obj2[item] +=1;
        else obj2[item] = 1;
    }
    
    cnt2 = Object.keys(obj2).length;
    
    const len = topping.length;
    for(let i=0; i < len-1; i++) {
        const item = topping[i];
        
        if(obj1[item]) obj1[item] +=1;
        else {
            obj1[item] = 1;
            cnt1 +=1;
        }
        
        obj2[item] -=1;
        
        if(obj2[item] ==0) {
            delete obj2.item;
            cnt2 -=1;
        }
        
        if(cnt1 === cnt2) answer +=1;
    }
    
    return answer;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    
    map<int, int> m1;
    map<int, int> m2;
    int cnt1 =0;
    int cnt2 =0;
    
    for(int& item : topping)
    {
        if(m2.find(item) != m2.end()) m2[item] += 1;
        else {
            m2.insert({item, 1});
            cnt2 +=1;
        }
    }
    
    int len = topping.size();
    for(int i=0; i<len-1; i++) 
    {
        int item = topping[i];
        
        if(m1.find(item) != m1.end()) m1[item] += 1;
        else 
        {
            m1.insert({item, 1});
            cnt1 +=1;
        }
        
        m2[item] -= 1;
        
        if(m2[item] ==0) 
        {
            m2.erase(item);
            cnt2 -=1;
        }
        
        if(cnt1 == cnt2) answer +=1;
    }
    
    return answer;
}